pyctbgui/utils/numpyWriter/npy_writer.py

```diff
@@ -0,0 +1,223 @@
+"""
+Wrapper to be able to append frames to a numpy file
+
+numpy header v1
+
+- 6bytes                         \x93NUMPY
+- 1 byte major version number    \x01
+- 1 byte minor version number    \x00
+- 2 bytes (unsigned short) HEADER_LEN length of header to follow
+- Header as an ASCII dict terminated by \n padded with space \x20 to make sure
+we get len(magic string) + 2 + len(length) + HEADER_LEN divisible with 64
+Allocate enough space to allow for the data to grow
+"""
+
+import ast
+import logging
+import os
+import zipfile
+from pathlib import Path
+
+import numpy as np
+
+
+class NumpyFileManager:
+    """
+    class used to read and write into .npy files that can't be loaded completely into memory
+
+    for read mode implements numpy like interface and file-like object function
+    """
+    magic_str = np.lib.format.magic(1, 0)
+    headerLength = np.uint16(128)
+    FSEEK_FILE_END = 2
+    BUFFER_MAX = 500
+
+    def __init__(
+        self,
+        file: str | Path | zipfile.ZipExtFile,
+        mode: str = 'r',
+        frameShape: tuple = None,
+        dtype=None,
+    ):
+        """
+        initiates a NumpyFileManager class for reading or writing bytes directly to/from a .npy file
+        @param file: path to the file to open or create
+        @param frameShape: shape of the frame ex: (5000,) for waveforms or (400,400) for image
+        @param dtype: type of the numpy array's header
+        @param mode: file open mode must be in 'rwx'
+        """
+        if mode not in 'rwx':
+            raise ValueError('file mode should be either r,w,x')
+
+        if isinstance(file, zipfile.ZipExtFile):
+            if mode != 'r':
+                raise ValueError('NumpyFileManager only supports read mode for zipfiles')
+        else:
+            if mode == 'x' and Path.is_file(Path(file)):
+                raise FileExistsError(f'file {file} exists while given mode is x')
+
+        self.dtype = np.dtype(dtype)  # in case we pass a type like np.float32
+        self.frameShape = frameShape
+        self.frameCount = 0
+        self.logger = logging.getLogger('NumpyFileManager')
+        self.cursorPosition = self.headerLength
+        self.mode = mode
+        newFile = (mode == 'w' or mode == 'x')
+
+        # if newFile frameShape and dtype should be present
+        if newFile:
+            assert frameShape is not None
+            assert dtype is not None
+            # create/clear the file with mode wb+
+            self.file = open(file, 'wb+')
+            self.updateHeader()
+            self.flush()
+
+        else:
+            # opens file for read/write and check if the header of the file corresponds to the given function
+            # arguments
+            if isinstance(file, zipfile.ZipExtFile):
+                self.file = file
+            else:
+                self.file = open(file, 'rb+')
```

Comment:
If the user passes 'r' as mode I don't think we should open the file in rb+ since this allows for updating the file. If we open with the correct flags it should give and error when trying to write. 

In case the user want to be able to update the file he/she should pass r+